addon-test-support/@ember/test-helpers/-internal/helper-hooks.ts

```diff
@@ -0,0 +1,64 @@
+type Hook = (...args: any[]) => void | Promise<void>;
+type HookUnregister = {
+  unregister: () => void;
+};
+
+const registeredHooks = new Map<string, Set<Hook>>();
+
+/**
+ * @private
+ * @param {string} helperName The name of the test helper in which to run the hook.
+ * @param {string} label A label to help identify the hook.
+ * @returns {string} The compound key for the helper.
+ */
+function getHelperKey(helperName: string, label: string) {
+  return `${helperName}:${label}`;
+}
+
+/**
+ * Registers a hook function to be run during the invocation of a test helper.
+ *
+ * @private
+ * @param {string} helperName The name of the test helper in which to run the hook.
+ * @param {string} label A label to help identify the hook. Built-in labels are `start` and `end`,
+ *                       designating the start of the helper invocation and the end.
+ * @param {Function} hook The hook function to run when the test helper is invoked.
+ * @returns {HookUnregister} An object containing an unregister function that will unregister
+ *                           the specific hook registered to the helper.
+ */
+export function registerHook(helperName: string, label: string, hook: Hook): HookUnregister {
+  let helperKey = getHelperKey(helperName, label);
+  let hooksForHelper = registeredHooks.get(helperKey);
+
+  if (hooksForHelper === undefined) {
+    hooksForHelper = new Set<Hook>();
+    registeredHooks.set(helperKey, hooksForHelper);
+  }
+
+  hooksForHelper.add(hook);
+
+  return {
+    unregister() {
+      hooksForHelper!.delete(hook);
+    },
+  };
+}
+
+/**
+ * Runs all hooks registered for a specific test helper.
+ *
+ * @private
+ * @param {string} helperName  The name of the test helper.
+ * @param {string} label A label to help identify the hook. Built-in labels are `start` and `end`,
+ *                       designating the start of the helper invocation and the end.
+ * @param {any[]} args Any arguments originally passed to the test helper.
+ * @returns {Promise<void>} A promise representing the serial invocation of the hooks.
+ */
+export function runHooks(helperName: string, label: string, ...args: any[]): Promise<void> {
+  let hooks = registeredHooks.get(getHelperKey(helperName, label)) || new Set<Hook>();
+  let promise = Promise.resolve();
+
+  hooks.forEach(hook => promise.then(() => hook(...args)));
```

Comment:
This leaks the promise at the moment, I think we need to change to:

```suggestion
  hooks.forEach(hook => promise = promise.then(() => hook(...args)));
```

Specifically, the original code would not allow a hook to return a promise and have `runHooks` properly wait for it to resolve. Can you add a test for this?

Reply:
Something like:

```js
  test('it can register a hook that returns a promise / has a delay', async function (assert) {
    let func = () => {
      assert.step('starting hook');

      return new Promise((resolve) => {
        setTimeout(() => {
          assert.step('resolving hook promise');
          resolve();
        }, 100);
      });
    };
    let hook = _registerHook('click', 'start', func);

    try {
      // it runs the hook
      assert.step('running hooks for click:start');
      await _runHooks('click', 'start');
      assert.step('hooks finished for click:start');

      assert.verifySteps([
        'running hooks for click:start',
        'starting hook',
        'resolving hook promise',
        'hooks finished for click:start',
      ]);
    } finally {
      hook.unregister();
    }
  });
```

Reply:
Good catch! I've added this.